File: 2022/5/day5.c
#include <stdio.h>
#include <string.h>

int main()
{
    FILE *file = fopen("input/day5_input.txt", "r");
    char *line;
    char stacks[3][50] = {0};

    int current_stack = 0;
    while (fgets(line, 13, file))
    {
        if (line[0] == '\n')
            break;
        printf("%s\n", line);
        for (int i = 0; i < strlen(line); i++)
        {
            if (line[i] == ' ')
            {
                current_stack++;
                i += 3;
            }
            else if (line[i] == '[')
            {
                for (int j = 0; j < 49; j++)
                    stacks[current_stack][j + 1] = stacks[current_stack][j];
                stacks[current_stack][0] = line[i + 1];
                current_stack++;
                i += 3;
            }
            if (i < 50 && line[i] == '\n')
                current_stack = 0;
        }
    }

    for (int i = 0; i < 3; i++)
    {
        printf("%s\n", stacks[i]);
    }
    
    printf("end\n");

    return 0;
}

File: 2022/5/day5.py
with open("input/day5_input.txt") as f:
    input = f.read().split("\n\n")

stacks = [[], [], [], [], [], [], [], [], []]
currstack = 0
i = 0
while i < len(input[0]):
    if input[0][i] == ' ':
        i += 3
        currstack += 1
    elif input[0][i] == '[':
        stacks[currstack].insert(0, input[0][i + 1])
        currstack += 1
        i += 3
    if i < len(input[0]) and input[0][i] == '\n':
        currstack = 0
    i += 1

instr = input[1].split("\n")

instr = [x.split() for x in instr]
instr = [[int(x[1]), int(x[3]), int(x[5])] for x in instr]

stacks2 = list()
for i in stacks:
    stacks2.append(list(i))


for i in instr:
    for j in range(i[0]):
        e = stacks2[i[1] - 1].pop()
        stacks2[i[2] - 1].append(e)
    
s = ""
for stack in stacks2:
    s += stack[-1]
print("Sol1:", s)

for i in instr:
    a = list()
    for j in range(i[0]):
        e = stacks[i[1] - 1].pop()
        a.append(e)
    for j in range(i[0]):
        e = a.pop()
        stacks[i[2] - 1].append(e)

s = ""
for stack in stacks:
    s += stack[-1]
print("Sol2:", s)